core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java

```diff
@@ -54,23 +63,54 @@
 
     private final CompletableFuture<List<Endpoint>> initialEndpointsFuture = new EventLoopCheckingFuture<>();
     private final AsyncCloseableSupport closeable = AsyncCloseableSupport.of(this::closeAsync);
+    private boolean allowEmptyEndpoints;
```

Comment:
Leaving a simple diff of my idea in case my suggestion was unclear 😅 

```
diff --git a/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
index 613426286..4a9b1241f 100644
--- a/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
+++ b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java
@@ -16,7 +16,6 @@
 package com.linecorp.armeria.client.endpoint;
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
-import static java.util.Objects.requireNonNull;
 
 import java.util.ArrayList;
 import java.util.Collections;
@@ -63,7 +62,7 @@ public class DynamicEndpointGroup extends AbstractEndpointGroup implements Liste
 
     private final CompletableFuture<List<Endpoint>> initialEndpointsFuture = new EventLoopCheckingFuture<>();
     private final AsyncCloseableSupport closeable = AsyncCloseableSupport.of(this::closeAsync);
-    private boolean allowEmptyEndpoints;
+    private final boolean allowEmptyEndpoints;
 
     /**
      * Creates a new empty instance, using {@link EndpointSelectionStrategy#weightedRoundRobin()}
@@ -79,7 +78,7 @@ public class DynamicEndpointGroup extends AbstractEndpointGroup implements Liste
      * @param selectionStrategy the {@link EndpointSelectionStrategy} of this {@link EndpointGroup}
      */
     public DynamicEndpointGroup(EndpointSelectionStrategy selectionStrategy) {
-        this.selectionStrategy = requireNonNull(selectionStrategy, "selectionStrategy");
+        this(selectionStrategy, false);
     }
 
     /**
@@ -88,8 +87,7 @@ public class DynamicEndpointGroup extends AbstractEndpointGroup implements Liste
      * @param allowEmptyEndpoints whether to allow an empty {@link Endpoint} list
      */
     protected DynamicEndpointGroup(boolean allowEmptyEndpoints) {
-        this();
-        this.allowEmptyEndpoints = allowEmptyEndpoints;
+        this(EndpointSelectionStrategy.weightedRoundRobin(), allowEmptyEndpoints);
     }
 
     /**
@@ -99,7 +97,7 @@ public class DynamicEndpointGroup extends AbstractEndpointGroup implements Liste
      * @param allowEmptyEndpoints whether to allow an empty {@link Endpoint} list
      */
     protected DynamicEndpointGroup(EndpointSelectionStrategy selectionStrategy, boolean allowEmptyEndpoints) {
-        this(selectionStrategy);
+        this.selectionStrategy = selectionStrategy;
         this.allowEmptyEndpoints = allowEmptyEndpoints;
     }
 
diff --git a/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
index a1ee6faa9..d5c6dfafa 100644
--- a/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
+++ b/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupBuilder.java
@@ -15,7 +15,8 @@
  */
 package com.linecorp.armeria.client.endpoint;
 
-import com.linecorp.armeria.common.annotation.Nullable;
+import static java.util.Objects.requireNonNull;
+
 import com.linecorp.armeria.common.annotation.UnstableApi;
 
 /**
@@ -25,8 +26,7 @@ import com.linecorp.armeria.common.annotation.UnstableApi;
 public final class DynamicEndpointGroupBuilder
         extends AbstractDynamicEndpointGroupBuilder {
 
-    @Nullable
-    private EndpointSelectionStrategy selectionStrategy;
+    private EndpointSelectionStrategy selectionStrategy = EndpointSelectionStrategy.weightedRoundRobin();
 
     DynamicEndpointGroupBuilder() {}
 
@@ -35,7 +35,7 @@ public final class DynamicEndpointGroupBuilder
      * If unspecified, {@link EndpointSelectionStrategy#weightedRoundRobin()} is used.
      */
     public DynamicEndpointGroupBuilder selectionStrategy(EndpointSelectionStrategy selectionStrategy) {
-        this.selectionStrategy = selectionStrategy;
+        this.selectionStrategy = requireNonNull(selectionStrategy, "selectionStrategy");
         return this;
     }
 
@@ -48,9 +48,6 @@ public final class DynamicEndpointGroupBuilder
      * Returns a newly created {@link DynamicEndpointGroup} with the properties configured so far.
      */
     public DynamicEndpointGroup build() {
-        if (selectionStrategy != null) {
-            return new DynamicEndpointGroup(selectionStrategy, shouldAllowEmptyEndpoints());
-        }
-        return new DynamicEndpointGroup(shouldAllowEmptyEndpoints());
+        return new DynamicEndpointGroup(selectionStrategy, shouldAllowEmptyEndpoints());
     }
 }

```

Reply:
Done